js/views/modals/purchase/Purchase.js

```diff
@@ -237,7 +294,11 @@ export default class extends BaseModal {
   changeCoupons(hashes, codes) {
     // combine the codes and hashes so the receipt can check both.
     // if this is the user's own listing they will have codes instead of hashes
-    this.receipt.coupons = hashes.concat(codes);
+    const hashesAndCodes = hashes.concat(codes);
+    const filteredCoupons = this.listing.get('coupons').filter((coupon) =>
+      hashesAndCodes.indexOf(coupon.get('hash') || coupon.get('discountCode')) !== -1);
+    this.couponObj = filteredCoupons.map(coupon => coupon.toJSON());
+    this.receipt.coupons = this.couponObj;
```

Comment:
This was moved from the receipt view to here, so the coupon data is available for the calculation of the current total.